subsys/net/lib/sockets/sockets.c

```diff
@@ -1539,6 +1634,205 @@ ssize_t z_vrfy_zsock_recvfrom(int sock, void *buf, size_t max_len, int flags,
 #include <syscalls/zsock_recvfrom_mrsh.c>
 #endif /* CONFIG_USERSPACE */
 
+ssize_t zsock_recvmsg_ctx(struct net_context *ctx, struct msghdr *msg,
+			  int flags)
+{
+	enum net_sock_type sock_type = net_context_get_type(ctx);
+	size_t i, max_len = 0;
+
+	if (msg == NULL) {
+		errno = -EINVAL;
+		return -1;
+	}
+
+	for (i = 0; i < msg->msg_iovlen; i++) {
+		max_len += msg->msg_iov[i].iov_len;
+	}
+
+	if (sock_type == SOCK_DGRAM) {
+		return zsock_recv_dgram(ctx, msg, NULL, max_len, flags,
+					msg->msg_name, &msg->msg_namelen);
+	} else if (sock_type == SOCK_STREAM) {
+		return zsock_recv_stream(ctx, msg, NULL, max_len, flags);
+	}
+
+	__ASSERT(0, "Unknown socket type");
+
+	errno = ENOTSUP;
+
+	return -1;
+}
+
+ssize_t z_impl_zsock_recvmsg(int sock, struct msghdr *msg, int flags)
+{
+	int bytes_received;
+
+	bytes_received = VTABLE_CALL(recvmsg, sock, msg, flags);
+
+	sock_obj_core_update_recv_stats(sock, bytes_received);
+
+	return bytes_received;
+}
+
+#ifdef CONFIG_USERSPACE
+ssize_t z_vrfy_zsock_recvmsg(int sock, struct msghdr *msg, int flags)
+{
+	struct msghdr msg_copy;
+	size_t iovlen;
+	size_t i;
+	int ret;
+
+	if (msg == NULL) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	K_OOPS(k_usermode_from_copy(&msg_copy, (void *)msg, sizeof(msg_copy)));
+
+	k_usermode_from_copy(&iovlen, &msg->msg_iovlen, sizeof(iovlen));
+
+	msg_copy.msg_name = NULL;
+	msg_copy.msg_control = NULL;
+
+	msg_copy.msg_iov = k_usermode_alloc_from_copy(msg->msg_iov,
+				       msg->msg_iovlen * sizeof(struct iovec));
+	if (!msg_copy.msg_iov) {
+		errno = ENOMEM;
+		goto fail;
+	}
+
+	for (i = 0; i < iovlen; i++) {
```

Comment:
Shall we have an initial loop run over allocated `msg_iov` entries, to clear `iov_base` pointer? Right now, in case `msg_iovlen > 1` and let's say allocation for the first `msg_iov` entry fails, we still have remaining `iov_base` pointer copies from the original msghdr. This will cause trouble when we jump to fail handler (will try to release the memory we did not allocate here).

Reply:
Yes, good point.